src/ipa_pc/data_structures.rs

```diff
@@ -51,83 +107,89 @@ impl<G: AffineCurve> PCUniversalParams for UniversalParams<G> {
 /// polynomial.
 #[derive(Derivative)]
 #[derivative(
-    Default(bound = ""),
     Hash(bound = ""),
     Clone(bound = ""),
     Debug(bound = ""),
     Eq(bound = ""),
     PartialEq(bound = "")
 )]
 #[derive(CanonicalSerialize, CanonicalDeserialize)]
-pub struct CommitterKey<G: AffineCurve> {
-    /// The key used to commit to polynomials.
-    pub comm_key: Vec<G>,
+pub struct CommitterKey<G: Curve> {
+    /// The base point vector used for the coefficients of the polynomial.
+    pub comm_key: Vec<G::AffineRep>,
 
-    /// A random group generator.
+    /// The base point used by the opening proof to reduce to a non-blinded
+    /// opening assertion.
     pub h: G,
 
-    /// A random group generator that is to be used to make
-    /// a commitment hiding.
+    /// The base point that is to be used for hiding.
     pub s: G,
 
     /// The maximum degree supported by the parameters
     /// this key was derived from.
     pub max_degree: usize,
 
-    /// H(max_degree_comm_key, h, s, max_degree)
+    /// A fingerprint of the CommitterKey, H(max_degree_comm_key, h, s, max_degree).
+    /// Used for reference.
     pub hash: Vec<u8>,
 }
 
-impl<G: AffineCurve> CommitterKey<G> {
+impl<G: Curve> CommitterKey<G> {
     /// Scale key for testing purpose
     pub fn scale(&mut self, scaling_scalar: G::ScalarField) {
-        self.h = self.h.mul(scaling_scalar).into_affine();
+        self.h = self.h.mul(&scaling_scalar);
     }
 }
 
-impl<G: AffineCurve> SemanticallyValid for CommitterKey<G> {
+impl<G: Curve> SemanticallyValid for CommitterKey<G> {
     // Technically this function is redundant, since the keys are generated
     // through a deterministic procedure starting from a public string.
     fn is_valid(&self) -> bool {
-        self.comm_key.is_valid()
+        G::batch_from_affine(&self.comm_key).is_valid()
             && self.h.is_valid()
             && self.s.is_valid()
-            && PCCommitterKey::supported_degree(self) <= self.max_degree
+            && PCCommitterKey::segment_size(self) <= self.max_degree
     }
 }
 
-impl<G: AffineCurve> PCCommitterKey for CommitterKey<G> {
+impl<G: Curve> PCCommitterKey for CommitterKey<G> {
     fn max_degree(&self) -> usize {
         self.max_degree
     }
-    fn supported_degree(&self) -> usize {
+
+    fn segment_size(&self) -> usize {
         self.comm_key.len() - 1
     }
 
     fn get_hash(&self) -> &[u8] {
         self.hash.as_slice()
```

Comment:
And let us clarify that we return the hash of the "master" key instead of the current instance, to allow our mainchain verifier using the single master key when verifying proofs that are produced with trimmed keys. As the used (trimmed) key length can be always read from an IPA proof (by counting the L and R's) it is secure to bind a SNARK proof to the identifier/hash of the master key alone.

Reply:
Done

Reply:
I resolve this comment here, and move it to the security discussion of the Fiat-Shamir rng init.